### ISSUE TITLE
Handle linear empty case

### ISSUE THREAD

**Issue by aspiwack** [tweag/ghc#25]
The following ought to typecheck according to the rules from the paper (and standard linear logic)

```haskell
f :: a ->. Void ->. k
f x v = case v of {}
```

**Comment by monoidal** [comment]
Two possible approaches are:

(1) Introduce the notion of a bottom usage that is neutral element of `sup`, return bottom in appropriate usage environments etc.
(2) Currently, in Core, case expressions remember their type so that we know the type of an empty case. Make them remember usage environment.

**Comment by aspiwack** [comment]
As far as I can tell, it currently works. With linear base:

```haskell
> :set -XLinearTypes 
> :set -XLambdaCase 
> :set -XEmptyCase 
> import Data.Void
> import Prelude.Linear

Prelude Data.Void Prelude.Linear> f :: a #-> Void #-> k; f x v = v & \case {}
```

Though I seem to remember a way to make a variant of this fail from the reddit thread. Even if there is no bug, we have vaguely formulated plans for case expressions to have an attached usage environment instead of the current solution, and we may have to revisit this issue specifically when that happens. I'm keeping this around, but until it's a confirmed, not moving upstream.

**Comment by monoidal** [comment]
Empty case works and we have a test (`LinearEmptyCase`). The one from reddit thread https://www.reddit.com/r/haskell/comments/hbzqsj/linear_types_are_merged_in_ghc/fviaxdo/ works too thanks to the fix for upstream:#18731.

The only remaining thing here is remembering usage environment in case, but this is covered by upstream:#18768. Closing.
